File: packages/react-native/android/src/main/java/com/voicevox/HybridVoicevox.kt
package com.voicevox

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import com.margelo.nitro.voicevox.DownloadResult
import com.margelo.nitro.voicevox.HybridVoicevoxSpec
import com.margelo.nitro.voicevox.VoicevoxCharacter
import com.margelo.nitro.voicevox.VoicevoxModelInfo
import com.margelo.nitro.voicevox.VoicevoxStyle
import jp.voicevox.android.Voicevox

/**
 * Nitro HybridObject 実装。
 * packages/android の [Voicevox] Facade を呼ぶだけの薄いグルー。
 */
class HybridVoicevox : HybridVoicevoxSpec() {
    private var voicevox: Voicevox? = null

    private fun requireVoicevox(): Voicevox =
        voicevox ?: throw IllegalStateException("call initialize() first")

    override fun initialize(maxConcurrentDownloads: Double): Promise<Unit> =
        Promise.async {
            val context = NitroModules.applicationContext
                ?: throw IllegalStateException("no application context")
            voicevox = Voicevox.create(
                context = context,
                maxConcurrentDownloads = maxConcurrentDownloads.toInt(),
            )
        }

    override fun listModels(): Promise<Array<VoicevoxModelInfo>> {
        val vv = requireVoicevox()
        return Promise.async {
            vv.listModels().map { it.toNitro() }.toTypedArray()
        }
    }

    override fun getTermsURL(): String = requireVoicevox().termsURL

    override fun acceptLicense(modelId: String) {
        requireVoicevox().acceptLicense(modelId)
    }

    override fun isLicenseAccepted(modelId: String): Boolean =
        requireVoicevox().isLicenseAccepted(modelId)

    override fun downloadModel(id: String): Promise<Unit> {
        val vv = requireVoicevox()
        return Promise.async { vv.downloadModel(id) }
    }

    override fun downloadModels(ids: Array<String>): Promise<Array<DownloadResult>> {
        val vv = requireVoicevox()
        return Promise.async {
            val results = vv.downloadModels(ids.toList())
            ids.map { id ->
                DownloadResult(
                    modelId = id,
                    error = results[id]?.exceptionOrNull()?.toString(),
                )
            }.toTypedArray()
        }
    }

    override fun downloadAllModels(): Promise<Array<DownloadResult>> {
        val vv = requireVoicevox()
        return Promise.async {
            vv.downloadAllModels().map { (id, result) ->
                DownloadResult(modelId = id, error = result.exceptionOrNull()?.toString())
            }.toTypedArray()
        }
    }

    override fun synthesis(text: String, modelId: String, styleId: Double?): Promise<ArrayBuffer> {
        val vv = requireVoicevox()
        return Promise.async {
            val wav = vv.synthesis(text, modelId, styleId?.toInt())
            ArrayBuffer.copy(java.nio.ByteBuffer.wrap(wav))
        }
    }
}

/** voicevox-core-android のモデル情報を Nitro 生成の構造体へ変換する。 */
private fun jp.voicevox.android.VoicevoxModelInfo.toNitro(): VoicevoxModelInfo =
    VoicevoxModelInfo(
        id = id,
        filename = filename,
        sizeBytes = sizeBytes.toDouble(),
        downloadURL = downloadURL,
        vvmId = vvmId,
        characters = characters.map { c ->
            VoicevoxCharacter(
                name = c.name,
                speakerUuid = c.speakerUuid,
                creditText = c.creditText,
                termsURL = c.termsURL,
                styles = c.styles.map { VoicevoxStyle(it.name, it.id.toDouble()) }
                    .toTypedArray(),
            )
        }.toTypedArray(),
        isDownloaded = isDownloaded,
    )

File: packages/react-native/ios/HybridVoicevox.swift
import AVFoundation
import Foundation
import NitroModules
// VoicevoxCore の Swift ソースは prepare-sources.sh により
// ios/VoicevoxCore/ にコピーされ、このターゲットへ直接コンパイルされる
// (別モジュールとしては存在しないため import 不要)。

/// Nitro HybridObject 実装。
/// packages/swift の `Voicevox` Facade を呼ぶだけの薄いグルー。
final class HybridVoicevox: HybridVoicevoxSpec {
    private var voicevox: Voicevox?
    private var player: AVAudioPlayer?

    private func requireVoicevox() throws -> Voicevox {
        guard let voicevox else {
            throw RuntimeError.error(withMessage: "call initialize() first")
        }
        return voicevox
    }

    func initialize(maxConcurrentDownloads: Double) throws -> Promise<Void> {
        Promise.async { [weak self] in
            let voicevox = try Voicevox(
                maxConcurrentDownloads: Int(maxConcurrentDownloads)
            )
            self?.voicevox = voicevox
        }
    }

    func listModels() throws -> Promise<[VoicevoxModel]> {
        let voicevox = try requireVoicevox()
        return Promise.async {
            await voicevox.listModels().map { $0.toNitro() }
        }
    }

    func getTermsURL() throws -> String {
        try requireVoicevox().termsURL
    }

    func acceptLicense(modelId: String) throws {
        try requireVoicevox().acceptLicense(modelId: modelId)
    }

    func isLicenseAccepted(modelId: String) throws -> Bool {
        try requireVoicevox().isLicenseAccepted(modelId: modelId)
    }

    func downloadModel(id: String) throws -> Promise<Void> {
        let voicevox = try requireVoicevox()
        return Promise.async {
            try await voicevox.downloadModel(id: id)
        }
    }

    func downloadModels(ids: [String]) throws -> Promise<[DownloadResult]> {
        let voicevox = try requireVoicevox()
        return Promise.async {
            let results = await voicevox.downloadModels(ids: ids)
            return ids.map { id in
                switch results[id] {
                case .success, .none:
                    return DownloadResult(modelId: id, error: nil)
                case .failure(let error):
                    return DownloadResult(modelId: id, error: String(describing: error))
                }
            }
        }
    }

    func downloadAllModels() throws -> Promise<[DownloadResult]> {
        let voicevox = try requireVoicevox()
        return Promise.async {
            let results = await voicevox.downloadAllModels()
            return results.map { id, result in
                switch result {
                case .success:
                    return DownloadResult(modelId: id, error: nil)
                case .failure(let error):
                    return DownloadResult(modelId: id, error: String(describing: error))
                }
            }
        }
    }

    func synthesis(text: String, modelId: String, styleId: Double?) throws -> Promise<ArrayBuffer> {
        let voicevox = try requireVoicevox()
        return Promise.async {
            let wav = try await voicevox.synthesis(
                text: text,
                modelId: modelId,
                styleId: styleId.map { UInt32($0) }
            )
            return try ArrayBuffer.copy(data: wav)
        }
    }

    func playWav(wav: ArrayBuffer) throws -> Promise<Void> {
        // ArrayBuffer は JS スレッド外では寿命保証がないため、先にコピーする
        let data = wav.toData(copyIfNeeded: true)
        return Promise.async { @MainActor in
            let player = try AVAudioPlayer(data: data)
            self.player?.stop()
            self.player = player
            player.play()
        }
    }

    func stopPlayback() throws {
        player?.stop()
        player = nil
    }
}

private extension VoicevoxModelInfo {
    /// VoicevoxCore のモデル情報を Nitro 生成の構造体(VoicevoxModel)へ変換する。
    func toNitro() -> VoicevoxModel {
        VoicevoxModel(
            id: id,
            filename: filename,
            sizeBytes: Double(sizeBytes),
            downloadURL: downloadURL,
            vvmId: vvmId,
            domains: domains,
            characters: characters.map { character in
                VoicevoxCharacter(
                    name: character.name,
                    speakerUuid: character.speakerUuid,
                    creditText: character.creditText,
                    termsURL: character.termsURL,
                    styles: character.styles.map {
                        VoicevoxStyle(name: $0.name, id: Double($0.id), type: $0.type)
                    }
                )
            },
            isDownloaded: isDownloaded
        )
    }
}
